diff --git a/drivers/amlogic/nand/aml_nand.c b/drivers/amlogic/nand/aml_nand.c
index 58550d8..b2212af 100755
--- a/drivers/amlogic/nand/aml_nand.c
+++ b/drivers/amlogic/nand/aml_nand.c
@@ -495,7 +495,7 @@ static int aml_nand_add_partition(struct aml_nand_chip *aml_chip)
 			mini_part_blk_num = 2;
 		else
 			mini_part_blk_num = (NAND_MINI_PART_SIZE >> phys_erase_shift);
-
+        start_blk = 0;
 		do {
 			offset = adjust_offset + start_blk * mtd->erasesize;
 			error = mtd->block_isbad(mtd, offset);
