diff --git a/drivers/amlogic/wifi/broadcm_40181/include/epivers.h b/drivers/amlogic/wifi/broadcm_40181/include/epivers.h
index 9ad51ae..07ba367 100755
--- a/drivers/amlogic/wifi/broadcm_40181/include/epivers.h
+++ b/drivers/amlogic/wifi/broadcm_40181/include/epivers.h
@@ -44,6 +44,6 @@
 #define EPI_VERSION_DEV		5.90.195
 
 
-#define	EPI_VERSION_STR		"5.90.195.26.1.3"
+#define	EPI_VERSION_STR		"5.90.195.26.1.6.1"
 
 #endif 
