diff --git a/src/com/amlogic/HdmiSwitch/HdmiSwitch.java b/src/com/amlogic/HdmiSwitch/HdmiSwitch.java
index ffec903..035e991 100755
--- a/src/com/amlogic/HdmiSwitch/HdmiSwitch.java
+++ b/src/com/amlogic/HdmiSwitch/HdmiSwitch.java
@@ -710,13 +710,15 @@ public class HdmiSwitch extends Activity {
             writeSysfs(VIDEO2_CTRL_PATH, "1");
 
             if (getCurMode().equals("720p")) {
-                writeSysfs(VIDEO2_FRAME_WIDTH_PATH, "500");
+                writeSysfs(VIDEO2_FRAME_WIDTH_PATH, "640");
             } else if (getCurMode().equals("1080p")) {
                 writeSysfs(VIDEO2_FRAME_WIDTH_PATH, "800");
             } else {
                 writeSysfs(VIDEO2_FRAME_WIDTH_PATH, "0");
             }
             
+            writeSysfs(VIDEO2_ZOOM_PATH, "105");
+            
             if (dualEnabled) {
                 writeSysfs(VIDEO2_SCREEN_MODE_PATH, "1");
                 writeSysfs(MODE_PATH_VOUT2, "null");
@@ -758,7 +760,7 @@ public class HdmiSwitch extends Activity {
     }    
     
     /** set osd blank*/    
-    private static int setFb0Blank(String blankStr) {
+    public static int setFb0Blank(String blankStr) {
     	//Log.i(TAG, "setFb0Blank: " + blankStr);
         File file = new File(FB0_BLANK_PATH);
         if (!file.exists()) {        	
